Poject capstone: Data Wrangling
Loan Defaulting data obtained from kaggle

In [1]:
#importing packages
import pandas as pd
import numpy as np

There are three different files that come with this data, the first dataset is of applicants that have yet to be approved. the second file is the description of the columns, some of the column names may be difficult to interperate so this is a key file thats needed in order to get a better understaning of our columns. the final file is the data for applications in the past which provide information as to why they did or did not get approved and other atributes about the applicants.


In [2]:
#loading all three files of the data set
application_data = pd.read_csv('LoanDefaultData/application_data.csv') #applications with no outcome
colDescription = pd.read_csv('LoanDefaultData/columns_description.csv') #description of the columns
previous_application = pd.read_csv('LoanDefaultData/previous_application.csv') #prior applications and their info.

In [3]:
#checking the dimensions of data to see if its complete.
print(np.shape(pd.DataFrame(colDescription)))
print(np.shape(pd.DataFrame(previous_application)))
print(np.shape(pd.DataFrame(application_data)))

(160, 5)
(1670214, 37)
(307511, 122)


as we can see, the column description file has 160 rows and 5 columns (actually, 161 X 5 since python starts with 0), printing the description file itself will tell us exactly what the file contains. it should contain the description of the columns in both previous_application and application_data. 

In [4]:
display(colDescription) #using 'display' as opposed to 'print' shows the data in a cleaner form

Unnamed: 0                     Table                        Row  \
0             1          application_data                 SK_ID_CURR   
1             2          application_data                     TARGET   
2             5          application_data         NAME_CONTRACT_TYPE   
3             6          application_data                CODE_GENDER   
4             7          application_data               FLAG_OWN_CAR   
..          ...                       ...                        ...   
155         209  previous_application.csv             DAYS_FIRST_DUE   
156         210  previous_application.csv  DAYS_LAST_DUE_1ST_VERSION   
157         211  previous_application.csv              DAYS_LAST_DUE   
158         212  previous_application.csv           DAYS_TERMINATION   
159         213  previous_application.csv  NFLAG_INSURED_ON_APPROVAL   

                                           Description  \
0                             ID of loan in our sample   
1    Target variable (1 - client with payment diffi...   
2          Identification if loan is cash or revolving   
3                                 Gender of the client   
4                        Flag if the client owns a car   
..                                                 ...   
155  Relative to application date of current applic...   
156  Relative to application date of current applic...   
157  Relative to application date of current applic...   
158  Relative to application date of current applic...   
159  Did the client requested insurance during the ...   

                                   Special  
0                                      NaN  
1                                      NaN  
2                                      NaN  
3                                      NaN  
4                                      NaN  
..                                     ...  
155  time only relative to the application  
156  time only relative to the application  
157  time only relative to the application  
158  time only relative to the application  
159                                    NaN  

[160 rows x 5 columns]

As we can see, the column description file does indeed tell us what the column names are for the other two files however, it would make more sence if we transpose the data in order to see all columns. we also have no need for the first column labeled as 'Unnamed: 0' so we can drop it before transposing the DF

In [5]:
colDescription.drop('Unnamed: 0', axis = 'columns', inplace = True)
colDescriptionT = colDescription.transpose()
display(colDescriptionT)

0    \
Table                application_data   
Row                        SK_ID_CURR   
Description  ID of loan in our sample   
Special                           NaN   

                                                           1    \
Table                                         application_data   
Row                                                     TARGET   
Description  Target variable (1 - client with payment diffi...   
Special                                                    NaN   

                                                     2    \
Table                                   application_data   
Row                                   NAME_CONTRACT_TYPE   
Description  Identification if loan is cash or revolving   
Special                                              NaN   

                              3                              4    \
Table            application_data               application_data   
Row                   CODE_GENDER                   FLAG_OWN_CAR   
Description  Gender of the client  Flag if the client owns a car   
Special                       NaN                            NaN   

                                             5    \
Table                           application_data   
Row                              FLAG_OWN_REALTY   
Description  Flag if client owns a house or flat   
Special                                      NaN   

                                           6                     7    \
Table                         application_data      application_data   
Row                               CNT_CHILDREN      AMT_INCOME_TOTAL   
Description  Number of children the client has  Income of the client   
Special                                    NaN                   NaN   

                                   8                 9    ...  \
Table                 application_data  application_data  ...   
Row                         AMT_CREDIT       AMT_ANNUITY  ...   
Description  Credit amount of the loan      Loan annuity  ...   
Special                            NaN               NaN  ...   

                                    150  \
Table          previous_application.csv   
Row                NAME_SELLER_INDUSTRY   
Description  The industry of the seller   
Special                             NaN   

                                                           151  \
Table                                 previous_application.csv   
Row                                                CNT_PAYMENT   
Description  Term of previous credit at application of the ...   
Special                                                    NaN   

                                                           152  \
Table                                 previous_application.csv   
Row                                           NAME_YIELD_GROUP   
Description  Grouped interest rate into small medium and hi...   
Special                                                grouped   

                                                           153  \
Table                                 previous_application.csv   
Row                                        PRODUCT_COMBINATION   
Description  Detailed product combination of the previous a...   
Special                                                    NaN   

                                                           154  \
Table                                 previous_application.csv   
Row                                         DAYS_FIRST_DRAWING   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                                                           155  \
Table                                 previous_application.csv   
Row                                             DAYS_FIRST_DUE   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                                    

looking better, now we need to reset the intex to reflect the column names. Once the column names are renamed properly, we can drop the 'Row' row of data frame since its not needed anymore. 

In [6]:
columns_desc = colDescription['Row'] 
colDescriptionT.rename((columns_desc), axis = 1, inplace = True)
colDescriptionT.drop('Row', axis = 'rows', inplace = True)

In [7]:
display(colDescriptionT)

SK_ID_CURR  \
Table                application_data   
Description  ID of loan in our sample   
Special                           NaN   

                                                        TARGET  \
Table                                         application_data   
Description  Target variable (1 - client with payment diffi...   
Special                                                    NaN   

                                      NAME_CONTRACT_TYPE  \
Table                                   application_data   
Description  Identification if loan is cash or revolving   
Special                                              NaN   

                      CODE_GENDER                   FLAG_OWN_CAR  \
Table            application_data               application_data   
Description  Gender of the client  Flag if the client owns a car   
Special                       NaN                            NaN   

                                 FLAG_OWN_REALTY  \
Table                           application_data   
Description  Flag if client owns a house or flat   
Special                                      NaN   

                                  CNT_CHILDREN      AMT_INCOME_TOTAL  \
Table                         application_data      application_data   
Description  Number of children the client has  Income of the client   
Special                                    NaN                   NaN   

                            AMT_CREDIT       AMT_ANNUITY  ...  \
Table                 application_data  application_data  ...   
Description  Credit amount of the loan      Loan annuity  ...   
Special                            NaN               NaN  ...   

                   NAME_SELLER_INDUSTRY  \
Table          previous_application.csv   
Description  The industry of the seller   
Special                             NaN   

                                                   CNT_PAYMENT  \
Table                                 previous_application.csv   
Description  Term of previous credit at application of the ...   
Special                                                    NaN   

                                              NAME_YIELD_GROUP  \
Table                                 previous_application.csv   
Description  Grouped interest rate into small medium and hi...   
Special                                                grouped   

                                           PRODUCT_COMBINATION  \
Table                                 previous_application.csv   
Description  Detailed product combination of the previous a...   
Special                                                    NaN   

                                            DAYS_FIRST_DRAWING  \
Table                                 previous_application.csv   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                                                DAYS_FIRST_DUE  \
Table                                 previous_application.csv   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                                     DAYS_LAST_DUE_1ST_VERSION  \
Table                                 previous_application.csv   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                                                 DAYS_LAST_DUE  \
Table                                 previous_application.csv   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                                              DAYS_TERMINATION  \
Table                                 previous_application.csv   
Description  Relative to application date of current applic...   
Special                  time only relative to the application   

                               

the following cells pring the last two files that come with this dataset. 
one file is called 'application_data' which contains atributes of current applicants while the second file called 'previous_application.csv' contains attributes of prior applicants and the outcome of thier loan i.e. did they get approved and did they default.

In [8]:
print('application_data DataFrame printed below')
display(application_data)
print('previous_applicaton DataFrame printed below')
display(previous_application)


application_data DataFrame printed below


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0           100002       1         Cash loans           M            N   
1           100003       0         Cash loans           F            N   
2           100004       0    Revolving loans           M            Y   
3           100006       0         Cash loans           F            N   
4           100007       0         Cash loans           M            N   
...            ...     ...                ...         ...          ...   
307506      456251       0         Cash loans           M            N   
307507      456252       0         Cash loans           F            N   
307508      456253       0         Cash loans           F            N   
307509      456254       1         Cash loans           F            N   
307510      456255       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
0                    Y             0          202500.0    406597.5   
1                    N             0          270000.0   1293502.5   
2                    Y             0           67500.0    135000.0   
3                    Y             0          135000.0    312682.5   
4                    Y             0          121500.0    513000.0   
...                ...           ...               ...         ...   
307506               N             0          157500.0    254700.0   
307507               Y             0           72000.0    269550.0   
307508               Y             0          153000.0    677664.0   
307509               Y             0          171000.0    370107.0   
307510               N             0          157500.0    675000.0   

        AMT_ANNUITY  ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20  \
0           24700.5  ...                 0                0                0   
1           35698.5  ...                 0                0                0   
2            6750.0  ...                 0                0                0   
3           29686.5  ...                 0                0                0   
4           21865.5  ...                 0                0                0   
...             ...  ...               ...              ...              ...   
307506      27558.0  ...                 0                0                0   
307507      12001.5  ...                 0                0                0   
307508      29979.0  ...                 0                0                0   
307509      20205.0  ...                 0                0                0   
307510      49117.5  ...                 0                0                0   

       FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                     0                        0.0                       0.0   
1                     0                        0.0                       0.0   
2                     0                        0.0                       0.0   
3                     0                        NaN                       NaN   
4                     0                        0.0                       0.0   
...                 ...                        ...                       ...   
307506                0                        NaN                       NaN   
307507                0                        NaN                       NaN   
307508                0                        1.0                       0.0   
307509                0                        0.0                       0.0   
307510                0                        0.0                       0.0   

        AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                              0.0                        0.0   
1                              0.0                        0.0   
2                              0.0                        0.0   
3                              NaN                        NaN   
4                              

previous_applicaton DataFrame printed below


SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  \
0           2030495      271877     Consumer loans     1730.430   
1           2802425      108129         Cash loans    25188.615   
2           2523466      122040         Cash loans    15060.735   
3           2819243      176158         Cash loans    47041.335   
4           1784265      202054         Cash loans    31924.395   
...             ...         ...                ...          ...   
1670209     2300464      352015     Consumer loans    14704.290   
1670210     2357031      334635     Consumer loans     6622.020   
1670211     2659632      249544     Consumer loans    11520.855   
1670212     2785582      400317         Cash loans    18821.520   
1670213     2418762      261212         Cash loans    16431.300   

         AMT_APPLICATION  AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  \
0                17145.0     17145.0               0.0          17145.0   
1               607500.0    679671.0               NaN         607500.0   
2               112500.0    136444.5               NaN         112500.0   
3               450000.0    470790.0               NaN         450000.0   
4               337500.0    404055.0               NaN         337500.0   
...                  ...         ...               ...              ...   
1670209         267295.5    311400.0               0.0         267295.5   
1670210          87750.0     64291.5           29250.0          87750.0   
1670211         105237.0    102523.5           10525.5         105237.0   
1670212         180000.0    191880.0               NaN         180000.0   
1670213         360000.0    360000.0               NaN         360000.0   

        WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  ...  \
0                         SATURDAY                       15  ...   
1                         THURSDAY                       11  ...   
2                          TUESDAY                       11  ...   
3                           MONDAY                        7  ...   
4                         THURSDAY                        9  ...   
...                            ...                      ...  ...   
1670209                  WEDNESDAY                       12  ...   
1670210                    TUESDAY                       15  ...   
1670211                     MONDAY                       12  ...   
1670212                  WEDNESDAY                        9  ...   
1670213                     SUNDAY                       10  ...   

         NAME_SELLER_INDUSTRY  CNT_PAYMENT  NAME_YIELD_GROUP  \
0                Connectivity         12.0            middle   
1                         XNA         36.0        low_action   
2                         XNA         12.0              high   
3                         XNA         12.0            middle   
4                         XNA         24.0              high   
...                       ...          ...               ...   
1670209             Furniture         30.0        low_normal   
1670210             Furniture         12.0            middle   
1670211  Consumer electronics         10.0        low_normal   
1670212                   XNA         12.0        low_normal   
1670213                   XNA         48.0            middle   

                 PRODUCT_COMBINATION  DAYS_FIRST_DRAWING DAYS_FIRST_DUE  \
0           POS mobile with interest            365243.0          -42.0   
1                   Cash X-Sell: low            365243.0         -134.0   
2                  Cash X-Sell: high            365243.0         -271.0   
3                Cash X-Sell: middle            365243.0         -482.0   
4                  Cash Street: high                 NaN            NaN   
...                              ...                 ...            ...   
1670209   POS industry with interest            365243.0         -508.0   
1670210   POS industry with interest            365243.0        -1604.0   
1670211  POS household with interest      

Comparing both of these files, we see that 'application_data' does not have the outcome of the loan meaning, we do not know if the approved applicants defaulted on their loan(s) thus we will focus our attention on the file 'previous_application.csv'

for the sake of getting a better visual idea of what the column contains, the following cells create a new list with the name of the columns, and their description. i then created a temporary data frame that can be refered to if needed

In [9]:
one = []
for col in previous_application.columns:
    one.append(col)

In [11]:
dfname = {'name':one, 'desc':two}
display(pd.DataFrame(dfname))

name  \
0                    SK_ID_PREV   
1                    SK_ID_CURR   
2            NAME_CONTRACT_TYPE   
3                   AMT_ANNUITY   
4               AMT_APPLICATION   
5                    AMT_CREDIT   
6              AMT_DOWN_PAYMENT   
7               AMT_GOODS_PRICE   
8    WEEKDAY_APPR_PROCESS_START   
9       HOUR_APPR_PROCESS_START   
10  FLAG_LAST_APPL_PER_CONTRACT   
11       NFLAG_LAST_APPL_IN_DAY   
12            RATE_DOWN_PAYMENT   
13        RATE_INTEREST_PRIMARY   
14     RATE_INTEREST_PRIVILEGED   
15       NAME_CASH_LOAN_PURPOSE   
16         NAME_CONTRACT_STATUS   
17                DAYS_DECISION   
18            NAME_PAYMENT_TYPE   
19           CODE_REJECT_REASON   
20              NAME_TYPE_SUITE   
21             NAME_CLIENT_TYPE   
22          NAME_GOODS_CATEGORY   
23               NAME_PORTFOLIO   
24            NAME_PRODUCT_TYPE   
25                 CHANNEL_TYPE   
26             SELLERPLACE_AREA   
27         NAME_SELLER_INDUSTRY   
28                  CNT_PAYMENT   
29             NAME_YIELD_GROUP   
30          PRODUCT_COMBINATION   
31           DAYS_FIRST_DRAWING   
32               DAYS_FIRST_DUE   
33    DAYS_LAST_DUE_1ST_VERSION   
34                DAYS_LAST_DUE   
35             DAYS_TERMINATION   
36    NFLAG_INSURED_ON_APPROVAL   

                                                 desc  
0   ID of previous credit in Home credit related t...  
1                            ID of loan in our sample  
2   Contract product type (Cash loan, consumer loa...  
3                     Annuity of previous application  
4   For how much credit did client ask on the prev...  
5   Final credit amount on the previous applicatio...  
6            Down payment on the previous application  
7   Goods price of good that client asked for (if ...  
8   On which day of the week did the client apply ...  
9   Approximately at what day hour did the client ...  
10  Flag if it was last application for the previo...  
11  Flag if the application was the last applicati...  
12                            Flag Micro finance loan  
13    Down payment rate normalized on previous credit  
14        Interest rate normalized on previous credit  
15        Interest rate normalized on previous credit  
16                           Purpose of the cash loan  
17  Contract status (approved, cancelled, ...) of ...  
18  Relative to current application when was the d...  
19  Payment method that client chose to pay for th...  
20          Why was the previous application rejected  
21  Who accompanied client when applying for the p...  
22  Was the client old or new client when applying...  
23  What kind of goods did the client apply for in...  
24  Was the previous application for CASH, POS, CA...  
25      Was the previous application x-sell o walk-in  
26  Through which channel we acquired the client o...  
27  Selling area of seller place of the previous a...  
28                         The industry of the seller  
29  Term of previous credit at application of the ...  
30  Grouped interest rate into small medium and hi...  
31  Detailed product combination of the previous a...  
32  Relative to application date of current applic...  
33  Relative to application date of current applic...  
34  Relative to application date of current applic...  
35  Relative to application date of current applic...  
36  Relative to application date of current applic...